### ISSUE TITLE
After updating to v3.11.1 rewrite/redirects append a port to the URL

### ISSUE THREAD

**Issue by pepijn-vanvlaanderen** [amannn/next-intl#987]
### Description

The last change for internal redirecting appends a port to the URL. However in our setup at Railway.app this does a faulty redirect. When the middleware should rewrite / redirect, it takes a long time and eventually redirects to the correct URL but with the port number and that crashes.

Reverting to v3.11.0 works.

### Verifications

- [X] I've verified that the problem I'm experiencing isn't covered in [the docs](https://next-intl-docs.vercel.app/).
- [X] I've searched for similar, existing issues on GitHub and [Stack Overflow](https://stackoverflow.com/search?q=next-intl).
- [X] I've compared my app to [a working example](https://next-intl-docs.vercel.app/examples) to look for differences.

### Mandatory reproduction URL

https://yeahgummy.webbers.dev/cart

### Reproduction description

Steps to reproduce:
1. Open reproduction
5. Wait about a minute or so
6. See error: the page does not work and is redirected to `https://yeahgummy.webbers.dev:3000/winkelwagen`

### Expected behaviour

A redirect to `https://yeahgummy.webbers.dev/winkelwagen` showing a page.

**Comment by awkaiser-tr** [comment]
@pepijn-vanvlaanderen [your suggestion of config to control what port is used](https://github.com/amannn/next-intl/issues/658#issuecomment-2043424613) may be the way to go, but it's also possible that the current logic is suboptimal. To troubleshoot further, what request headers are available to Next.js? Do you have values for any of the `x-forwarded-*` headers, e.g.

- `x-forwarded-host`
- `x-forwarded-proto`
- `x-forwarded-port`

**Comment by amannn** [comment]
@pepijn-vanvlaanderen Sorry for the inconvenience caused by this change!

But yes, as @awkaiser-tr has pointed out, it would be important to get details about the specific request headers your app receives. Based on that, we can add a failing test like @awkaiser-tr did in https://github.com/amannn/next-intl/pull/979 and look into an appropriate fix.

I'd really like to avoid config as much as possible if we can, therefore the best option from my perspective would be to just do the right thing based on the information we have available.

Can you also share your `domains` config?

@awkaiser-tr Can you too share your `domains` config? Maybe we could also utilize a port that is potentially defined there for the detection.

**Comment by awkaiser-tr** [comment]
My `domains` config looks like:
```tsx
{
  defaultLocale: 'de-DE',
  domain: getDomain('de-DE').host,
  locales: ['de-DE'],
},
{
  defaultLocale: 'en-CA',
  domain: getDomain('en-CA').host,
  locales: ['en-CA', 'fr-CA'],
},
{
  defaultLocale: 'en-GB',
  domain: getDomain('en-GB').host,
  locales: ['en-GB'],
},
// ... and so on ...
```
... where `host` is the locale domain for current Next.js environment, e.g.

- `development` @ `de-yourcompany-dev.localhost`
- `production` @ `www.yourcompany.de`

**Comment by awkaiser-tr** [comment]
@amannn are you suggesting an optional `port` property on the per-domain config?

Since most deployments will be reached via standard HTTP (80) / HTTPS (443) port, treating non-standard ports as outliers that require config makes sense and would probably result in fewer issues filed on this project.[^1]

However, if `DomainConfig` were to begin caring about `port`, it would also need to consider `protocol`:
```tsx
{ // For sake of discussion ...
  defaultLocale: 'en-CA',
  domain: getDomain('en-CA').host, // `'www.yourcompany.ca'` || `'ca-yourcompany-dev.localhost'`
  locales: ['en-CA', 'fr-CA'],
  protocol: getDomain('en-CA').protocol, // `'https://'` || `'http://'`
  port: getDomain('en-CA').port, // `undefined` || `3000`
},
```
... where by default `next-intl` redirects use Next.js' request `protocol` and omit `port`, overridden by `DomainConfig` values first and proxy header values second.

[^1]: Do not many people test domain-based l10n during development on alternative ports like Next.js' default dev port `3000`? 🤔

**Comment by amannn** [comment]
Yep, that's correct. We currently use the protocol and port of an incoming request to determine wether to add it to a domain that we're redirecting to.

I was wondering if the port could even be part of `domain` (e.g. `domain: 'www.yourcompany.de:3000'`) to avoid introducing new properties, but I guess that would be somewhat of a stretch for the term "domain".

A separate property like you're suggesting might be the most explicit option and also gives users more flexibility (maybe one domain runs on a different port than another domain). I think this might be a reasonable path forward without having to rely on detecting a reverse proxy, which seemingly people have different configurations for.

What do you think?

**Comment by pepijn-vanvlaanderen** [comment]
Hi @amannn , hereby my domains config:
```
    localePrefix: 'as-needed',
    localeDetection: false,
    locales: ['nl'],
    defaultLocale: 'nl',
    pathnames: {
        ...,
        '/cart': {
            en: '/cart',
            nl: '/winkelwagen',
        },
        ...
    },
    domains: [
        {
            domain: `${process.env.DOMAIN_NL ?? 'localhost:3000'}`,
            defaultLocale: 'nl',
        },
    ],
```

As you can see, we currently only support 1 language as we are still waiting on translated content, but internally we use `next-intl` already to translate short texts.

@awkaiser-tr how does your `getDomain` look like? I was as well wondering for an easy ussage on local developing here.

**Comment by awkaiser-tr** [comment]
# Local development w/ locale-based domains

@pepijn-vanvlaanderen the `getDomain` function returns `{ host, origin, pathname, port, protocol }` by `locale`, with values determined by environment. You just need to associate some locale-specific domains with your loopback address, e.g. **/etc/hosts**:
```
127.0.0.1 ca-yourcompany-dev.localhost
127.0.0.1 de-yourcompany-dev.localhost
127.0.0.1 es-yourcompany-dev.localhost
127.0.0.1 fr-yourcompany-dev.localhost
127.0.0.1 it-yourcompany-dev.localhost
127.0.0.1 pt-yourcompany-dev.localhost
127.0.0.1 uk-yourcompany-dev.localhost
127.0.0.1 us-yourcompany-dev.localhost
127.0.0.1 yourcompany-dev.localhost
```

**Comment by awkaiser-tr** [comment]
> @amannn: I was wondering if the port could even be part of `domain` (e.g. `domain: 'www.yourcompany.de:3000'`) to avoid introducing new properties, but I guess that would be somewhat of a stretch for the term "domain".

Agreed, that would be overloading `domain` in a surprising way. 🙅🏻

> ~A separate property like you're suggesting might be the most explicit option.~ **EDIT**: Seems like we've silently already supported this, so maybe this is the way to go and we should document it.
> 
> An explicit port in the domain config should also give users more flexibility, maybe one domain runs on a different port than another domain. I think this might be a reasonable path forward without having to rely on detecting a reverse proxy, which seemingly people have different configurations for.
> 
> What do you think?

Adding `port` to `DomainConfig` would be fine, so long as `protocol` (e.g. `https://`, `http://`) is also available. A secure or insecure web server can listen on any available port, and the protocol of an internal Next.js server could mismatch its public-facing reverse proxy, so we can't make assumptions about `protocol` with no way to override. Presumably this also means consideration of proxy-related headers would be removed? Leaving it up to the user, with sensible defaults in absence of `DomainConfig` values, does feel like the best choice. If they want to set config values based on incoming headers, they'd be free to do that.

**Comment by pepijn-vanvlaanderen** [comment]
Hi @amannn, after some testing I see now that the issue occurs on the pathname rewrite. So i.e. I have in pathnames this:
```
pathnames: {
    '/cart': : {
        de: '/cart',
        en: '/cart',
        fr: '/cart',
        it: '/cart', 
        nl: '/winkelwagen',
    },
}
```
For all (custom) domains it will work, except for the nl (Dutch) variant. It will take some long time loading and end up at 
`https://yeahgummy-de.webbers.dev/cart` -> `https://yeahgummy-de.webbers.dev/cart`
`https://yeahgummy-fr.webbers.dev/cart` -> `https://yeahgummy-fr.webbers.dev/cart`
`https://yeahgummy-it.webbers.dev/cart` -> `https://yeahgummy-it.webbers.dev/cart`
`https://yeahgummy-nl.webbers.dev/cart` -> `https://yeahgummy-nl.webbers.dev:3000/winkelmand`

When logging the x-forwarded-host and x-forwarded-port on Railway, I get these results:
```
xForwardedHost: yeahgummy-fr.webbers.dev
xForwardedPort: 3000
```

**Comment by awkaiser-tr** [comment]
@amannn funny enough, I just experienced the same bug RE: localized pathname redirects 💣 

@pepijn-vanvlaanderen quick sanity check: I see you're reporting `x-forwarded-port` as `3000`, but was that just for the `https://yeahgummy-nl.webbers.dev:3000/winkelmand` request? What is the value of the `x-forwarded-*` headers for the `/cart` requests?

**Comment by pepijn-vanvlaanderen** [comment]
@awkaiser-tr for all the requests the port is set as 3000 and the domain I have set per locale as the host.

**Comment by TomasLukes** [comment]
> @awkaiser-tr for all the requests the port is set as 3000 and the domain I have set per locale as the host.

Same problem with us, when updating to 3.11.1

**Comment by amannn** [comment]
@pepijn-vanvlaanderen Are you saying that for a request to `https://yeahgummy-nl.webbers.dev/cart`, you're seeing `x-forwarded-port: 3000`? If so, that sounds like a bug of the reverse proxy to me, `x-forwarded-port` should contain the value of the port of the original request (443; standard https port).

See also:

> Load Balancing with Port Preservation: In a load-balanced environment (NGINX) where the backend servers are listening on different ports, the X-Forwarded-Port header helps the load balancer preserve and convey the original client port information to the server. This information can be used by the server for various purposes, such as generating absolute URLs.

([source](https://requestly.com/blog/what-are-x-forwarded-headers-and-why-it-is-used/))

Can you verify this with Railway?

@awkaiser-tr Does this match your understanding too?

@TomasLukes Which hosting provider are you using? Can you also double check what the value of `x-forwarded-port` is when `https://something.pl/complaints` hits your middleware? The correct value should be 443.

**Comment by pepijn-vanvlaanderen** [comment]
Hi @amannn, thanks for your message. Actually I think the issue then is probably how I had setup the Dockerfile. There is a reverse proxy in front of the container and inside the container it then uses the mentioned port. Maybe Railway should superseed that from the proxy, but in the meantime I was just able to expose port 443 in the dockerfile instead of 3000 and it works now, thanks!

**Comment by awkaiser-tr** [comment]
> @amannn: Does this match your understanding too?

Yup.

----

@pepijn-vanvlaanderen when you say "it works now" do you mean that, in addition to your other routes, the `https://yeahgummy-nl.webbers.dev/cart` localized pathname redirect also now leads to `https://yeahgummy-nl.webbers.dev/winkelmand` as expected?

This GitHub Issue has been confused since details have changed over time:

1. ~Undesired port _always_ added to redirects~
2. Specifically, localized pathname redirects include undesired port

**Comment by TomasLukes** [comment]
> @TomasLukes Which hosting provider are you using? Can you also double check what the value of `x-forwarded-port` is when `https://something.pl/complaints` hits your middleware? The correct value should be 443.

Thanks for quick help!👌 "x-forwarded-port" value in middleware is 3000 now. Should I do something like this?
`const modifiedReq = req.headers.set("x-forwarded-port", "443");`

It looks like its behaving correctly after that 🤔

**Comment by awkaiser-tr** [comment]
@TomasLukes [just as Pepijn did,](https://github.com/amannn/next-intl/issues/987#issuecomment-2085049111) I'd recommend digging deeper and determining why the `x-forwarded-port` header is being set to `3000` and correcting that _before_ the request reaches Next.js

**Comment by pepijn-vanvlaanderen** [comment]
@awkaiser-tr yes the redirect also leads to localized pathnames correctly. The undesired port issue originated from a bad config. 

I exposed port number 3000 in the production environment before, and thats why it redirected to the domain with port addition after reading `x-forwarded-port` . So in the end I set the port in Railway to 443, which resulted in a proper config and working redirects. Thanks for the help!

**Comment by maps82** [comment]
I run my next application in docker and in the docker container it runs on 3000, but it is exposed to port 8000 on my host. On redirect from / to /de, the port 8000 is changed to 3000. I this expected for my scenario? After all docker just redirects the tcp traffic and is not an HTTP proxy that sets the x-forwared-for ports.

**Comment by martijnmichel** [comment]
same issue on latest and multi-domain setup :( and i can't seem to fix it as well by downgrading to 3.10, my site now always redirects to :4001 when no /locale is given, very bad

**Comment by martijnmichel** [comment]
right. I can confirm that emptying the x-forwarded-port solves the issue. though it feels quite hacky.

<img width="670" alt="Screenshot 2024-08-03 at 14 38 13" src="https://github.com/user-attachments/assets/a65a6cac-44d6-4f0e-9b12-18f5f8f413bc">
